### PR TITLE
Implment Workload Discovery Service for ztunnel

### PR DESCRIPTION
**Pull request by howardjohn**
This implements the very experimental WDS to serve as the control plane for https://github.com/istio/ztunnel

### CODE REVIEW

File: pilot/cmd/pilot-agent/options/options.go
Comment (costinm) [inline pilot/cmd/pilot-agent/options/options.go:139]: What changed ? We're still using agent with ztunnel ?

File: pilot/pkg/features/pilot.go
Comment (costinm) [inline pilot/pkg/features/pilot.go:588]: Not sure what this means...

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (costinm) [inline pilot/pkg/serviceregistry/aggregate/controller.go:58]: Ok for now - but I think the reverse map from IP to workload info ( and from VIP to service info ) should be core in the model, and updated when the pod is changing ( without iterating trough all registries/pods/etc). 

Also I would add a TODO to include the network name somewhere.

File: pilot/pkg/ambient/ambient.go
Comment (kyessenov) [reply pilot/pkg/ambient/ambient.go:58]: Agree with @costinm . https://github.com/istio/istio/issues/41527. There's a lot of value in adopting this for sidecar, that's what the original intent was, to optimize sidecar MX, automTLS with ORIGINAL_DST, etc.

**Comment by kyessenov** [comment]
How are we testing this against Envoy ambient filters?

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (howardjohn) [reply pilot/pkg/ambient/controller/workloadcache.go:63]: Ha, wasn't sure if I should include this or not... I ended up putting it behind a feature flag but can remove entirely if we want. The intent is it makes it really easy to run ztunnel on local machine:

* Setup redirect rule to 15001 (I do it on a special uid)
* Run whatever app you want (say `fortio server`)
* Then `sudo -u redirected curl localhost:8080` goes through the ztunnel, does HBONE to itself

Basically we register `localhost` as a workload so we know to send HBONE to it

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (howardjohn) [reply pilot/pkg/ambient/controller/workloadcache.go:72]: Its not actually a new watch/informer, this is using a shared informer. So just a custom event handler.

But I do agree. I had some vague ideas on improving the whole pipeline here even more, but nothing concrete

**Comment by howardjohn** [comment]
> How are we testing this against Envoy ambient filters?

This is not used for envoy, it is for https://github.com/istio/ztunnel

**Comment by kyessenov** [comment]
> > How are we testing this against Envoy ambient filters?
> 
> This is not used for envoy, it is for https://github.com/istio/ztunnel

WorkloadMetadata is used in Envoy, it's the core of new stats filter and MX for sidecars. Perhaps, we need to define a proto somewhere else?

**Comment by howardjohn** [comment]
> > > How are we testing this against Envoy ambient filters?
> > 
> > 
> > This is not used for envoy, it is for https://github.com/istio/ztunnel
> 
> WorkloadMetadata is used in Envoy, it's the core of new stats filter and MX for sidecars. Perhaps, we need to define a proto somewhere else?

Its not the same one and has different requirements. If those end up converging we can use the same proto, but one of the main points of the project is to be able to have exactly the info we need (nothing more, nothing less) in an efficient representation, instead of being tied to what Envoy uses.

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (stevenctl) [inline pilot/pkg/ambient/controller/workloadcache.go:110]: Why only for update? 

Should we also trigger for transition from `valid -> invalid`?

**Comment by kyessenov** [comment]
> > > > How are we testing this against Envoy ambient filters?
> > > 
> > > 
> > > This is not used for envoy, it is for https://github.com/istio/ztunnel
> > 
> > 
> > WorkloadMetadata is used in Envoy, it's the core of new stats filter and MX for sidecars. Perhaps, we need to define a proto somewhere else?
> 
> Its not the same one and has different requirements. If those end up converging we can use the same proto, but one of the main points of the project is to be able to have exactly the info we need (nothing more, nothing less) in an efficient representation, instead of being tied to what Envoy uses.

If it's not the same one, please don't use the same proto then? Where is the source of truth for the proto? All you're saying is applicable to Envoy, we can change Envoy version easily to adapt.

File: pkg/workloadapi/workload.proto
Comment (kyessenov) [inline pkg/workloadapi/workload.proto:20]: Can we define this proto in a standard repo, api, for example? Or at least give it some thought since it's clearly in a draft state.

File: pilot/pkg/xds/workload.go
Comment (stevenctl) [inline pilot/pkg/xds/workload.go:39]: I don't see calls to `AmbientWorkloads` anywhere. Are the changes to workloadcache relevant? I'm not opposed to doing away with it and converging on a single "kube controller".

File: pkg/workloadapi/workload.proto
Comment (kyessenov) [inline pkg/workloadapi/workload.proto:52]: cc @costinm This looks like a dependency on services as well, not just endpoints.

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:20]: We discussed this for about 15min in the working group - it is a draft. This is being merged into the experimental branch, not master. Once it's not a draft it will be in api

**Comment by howardjohn** [comment]
> > > > > How are we testing this against Envoy ambient filters?
> > > > 
> > > > 
> > > > This is not used for envoy, it is for https://github.com/istio/ztunnel
> > > 
> > > 
> > > WorkloadMetadata is used in Envoy, it's the core of new stats filter and MX for sidecars. Perhaps, we need to define a proto somewhere else?
> > 
> > 
> > Its not the same one and has different requirements. If those end up converging we can use the same proto, but one of the main points of the project is to be able to have exactly the info we need (nothing more, nothing less) in an efficient representation, instead of being tied to what Envoy uses.
> 
> If it's not the same one, please don't use the same proto then? Where is the source of truth for the proto? All you're saying is applicable to Envoy, we can change Envoy version easily to adapt.

it's not the same proto. It's entirely different. The source of truth is in this PR in the .proto file. We can rename it if there is confusion

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (costinm) [reply pilot/pkg/ambient/controller/workloadcache.go:63]: Sounds good, but not in istiod ? As I mentioned in the chat, I think ztunnel should support loading workoad configs from a file - or from a localhost Istiod. 

Any reason to stop using pilot-agent ? We still probably  want DNS, integration with other CA providers ( assuming they 
support same checks as Istiod to allow getting certs for pods on same node ) and all the other stuff.  
Would also simplify the code in ztunnel - and istiod too since we can isolate ztunnel dev/testing from Istiod,

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (costinm) [reply pilot/pkg/ambient/controller/workloadcache.go:72]: I can do a draft on master - I would like to also add the pod labels to the result, no need to send them on each connection or request unsecured if we can just get them from the source. I think this reverse map of (network, IP) -> (workload info) should be as efficient as possible, in particular since we want very low latency when pods go up and down. 

The existing pipeline in Istiod is optimized for large batch updates - this is a very low cost / low latency call we can handle
in near real time, we can likely bypass debounce and a lot of other mechanisms intended for huge configs distributed to all pods. 

I would assume ztunnel caches the result for some time and tries again - but will only ask for  pods for which active
connections are getting open, so a change will only need to be distributed to the smaller list of nodes that are 
actively using the pod. If a pod nobody is connected to is changed - no push will be sent. And the update will
be identically sent to all the nodes that need it. This can't use the old flow.

Comment (costinm) [reply pilot/pkg/ambient/controller/workloadcache.go:72]: BTW - my interest around this is also for the 'trusted network' model, where info about IP is used to derive the identity instead of using the cert. I suspect will also provide a better solution for telemetry.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto:52]: That's fine - it's info about the endpoints, we're not trying to normalize a database. 

What I would like in the comments is what the field is used for, and if we can improve the representation to reduce 
object allocs/size/etc - since the goal is to have an optimized case-specific proto, not to reuse abstract/generic protos. 

For example do we need the port list ? Do we need the fully qualified service or just the base name ( services can't select cross ns ) ? Do we want the service VIP as well ( or what does virtual ips mean ? )

**Review by costinm — COMMENTED** [review]
I hate bikesheding on names, but I kind of agree with Kuat, using Workload for the proto name may be confusing and is a big land grab. 

ZTunnelIPInfo, or IPInfo or ZTunnelPTR ( nod to DNS PTR records that map IP to name) or something less generic than Workload. 

Can be done in a separate PR or before merging to master - not a reason to block this PR.

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:21]: See my comments on the doc. What is the name of a workload ? Is it fully qualified, pod name only, WorkloadEntry name, etc. 

BTW - do we use the same thing for VIPs ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:28]: can it be an enum ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:30]: How is it different from name ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:36]: list, bytes ( I don't think we need decimal address )

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:39]: ServiceAccountName instead. We can add a list of identities for extras ( to support the APIs or for VIPs). No need to duplicate the namespace and trust domain, we know it already. K8S Pod also has just ServiceAccount.

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:45]: I don't know if we need this. If the pod needs to be reached via gateway  - just send the list of address ( and it may be more than one ). If the pod is reached via waypoint - that can also go into address field.

Comment (costinm) [reply pkg/workloadapi/workload.proto:45]: I don't mind too much keeping them separated - but like for address, please list ( dual stack, multi-nic), and I would stop the decimal and ascii address.

Also please add the 'network' field.

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:48]: When is this used ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:55]: Where is it used ? What is it ? Do we really need it if we have the pod name ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:58]: Not sure about this. Can we add it later, after we figure out how it'll work on trusted networks ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:61]: FQDN ? IP ? Why ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:62]: I would also add a labels field - with docs stating can be a subset of pod labels or the full set. 

Subset may be based on labels actively used in policies or telemetry.

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:67]: Same, not sure what is the use case for not enforcing, but I would like a design doc before we do another 'permisive'.

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:72]: I am not sure at all that the full policy in affect for an IP should be sent along with the info. 

This is a case where zproxy can get policies separately, and just have a reference to it. 

However - policy enforcement, at least in NetworkPolicy case, is based on label matching. I would not reinvent this - if the purpose is to reimplement 
NetworkPolicy ( since we are bypassing it ), just use NetworkPolicy and send the labels.

Comment (kyessenov) [reply pkg/workloadapi/workload.proto:72]: Cilium already has xDS server and schema for NetworkPolicies, which I assume is widely deployed.

Comment (zirain) [reply pkg/workloadapi/workload.proto:72]: > Cilium already has xDS server and schema for NetworkPolicies, which I assume is widely deployed.

can you share the link for that

Comment (kyessenov) [reply pkg/workloadapi/workload.proto:72]: https://github.com/cilium/proxy/blob/master/cilium/api/npds.proto. Implementation is in the same repo.

Comment (kyessenov) [reply pkg/workloadapi/workload.proto:72]: @zirain https://github.com/cilium/proxy/blob/master/cilium/network_policy.cc#L749 is the main xDS method, with a standard pattern to post to thread locals for every update. It is a SotW which seems to be enough for all Cilium clusters.

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:48]: This was basically for the case when the workload supports HBONE directly (sidecar, proxyless, etc), so we can make sure we pass through instead of terminating it. Might be a better way to do this, though.

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:52]: We only have the IP, no names of anything. I update the docs to be more clear.

Port list - we might be able to get rid of it in some cases. We do need to know if there is service_port!=target_port though.

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:55]: We used to, not anymore. Removed

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:61]: Name of the node. We need it since the ztunnel has different logic for node-local pod and remote pods.

Alternative would be this could just be a `bool` or something, but one nice thing is right now `Workload` doesn't depend on the client at all, so its very cacheable

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:62]: What would we use it for?

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:67]: If they have no policies

File: pilot/pkg/ambient/controller/workloadcache.go
Comment (costinm) [reply pilot/pkg/ambient/controller/workloadcache.go:58]: Agreed. Let's get this PR merged so ztunnel testing can proceed with the branch, and we can work on master - I think we can support both 'send all' and on-demand, and the set of data is almost the same for all cases.

File: pilot/pkg/serviceregistry/aggregate/controller.go
Comment (costinm) [reply pilot/pkg/serviceregistry/aggregate/controller.go:58]: I initially thought rust was using on-demand already. some of my comments were based on that. If we send all endpoints - this is ok.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto]: I think we need a lot more documentation - for each field, where is it used and what it can be.

We only have IP - but do we send mixed Services VIPs and Pod IPs (and external IPs from ServiceEntry ) ? Any case where we'll let k8s do the LB and the VIP will be used ? I'm not entirely sure what you capture in ztunnel, I assume it's primarily service VIPs ?  

Since we're dumping all IPs - it is fine to mix them I think. Pod IPs have the list of service names, services IPs have the service as well - ztunnel can join them.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto:28]: We really need to have a design doc and review for what telemetry we must collect for ambient, and how much cardinality and costs are tolerable and useful. 

IMO pod name is important for debugging and also may be useful for security and other purpose - I would keep it. 
Deployment and CronJob ? Do we really need to create timeseries for info the user can just do a kubectl to find out ?

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto]: Sorry - I meant "add to the docs what it is and an example and why" - I'm happy either way.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto:73]: NetworkPolicy implementation ? Since ztunnel is breaking the platform implementation, it'll likely need at some point to fix it and possibly have extended rules.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto]: Confused. What policies do they need ? My understanding was that mTLS is on by default and enforced, even for pod-to-pod. With CNI you can't opt out of wireguard/ipsec.

File: pkg/workloadapi/workload.proto
Comment (costinm) [reply pkg/workloadapi/workload.proto:90]: is it the same as k8s NetworkPolicy, or just another policy using same name ? It seems everything is now using same generic names and are 'standards' because they are sent over XDS,  even if they are very vendor specific.

Comment (kyessenov) [reply pkg/workloadapi/workload.proto:90]: Yes, looks like it's L7 network policy in Cilium. My point still stands, stop calling this "Policy", that's super confusing. Everything is a policy on some workloads.

Comment (costinm) [reply pkg/workloadapi/workload.proto:90]: Agreed, we should avoid unqualified broad names. Can't change Cilium name, but we should. 

I think Cilium validates the pattern of sending a custom vendor specific object over XDS - but I would postpone merging any authz related in the proto ( I assume it is not implemented or use in ztunnel right now - most of the fields don't seem to be ). The main question is if ztunnel will implement K8SNetworkPolicy ( with label-based extensions ), subset of Istio Auths, adopt some other standard - or invent yet another one. 

All this is not related to this PR.

**Review by costinm — COMMENTED** [review]
John, to simplify the review of this PR - can you remove all the fields from the proto that are not currently used in the rust implementation, and add a top comment in the proto that it's internally matching ztunnel rust and the 'real' proto will be reviewed in a standalone PR or several incremental PRs.

**Comment by howardjohn** [comment]
> John, to simplify the review of this PR - can you remove all the fields from the proto that are not currently used in the rust implementation, and add a top comment in the proto that it's internally matching ztunnel rust and the 'real' proto will be reviewed in a standalone PR or several incremental PRs.

Sure, it's just RBAC and the telemetry stuff currently not used. They were used in the Go version though

**Comment by costinm** [comment]
> > John, to simplify the review of this PR - can you remove all the fields from the proto that are not currently used in the rust implementation, and add a top comment in the proto that it's internally matching ztunnel rust and the 'real' proto will be reviewed in a standalone PR or several incremental PRs.
> 
> Sure, it's just RBAC and the telemetry stuff currently not used. They were used in the Go version though

Which RBAC ? Doesn't look like Istio RBAC, nor k8s RBAC ( which is full of labels ). That's why having qualified names is better...

**Comment by howardjohn** [comment]
The l4 subset of Istio rbac

File: pkg/workloadapi/workload.proto
Comment (kyessenov) [reply pkg/workloadapi/workload.proto:90]: I don't think it's right to project policy onto an endpoint in any case. It's the wrong delta of change. A policy changes should not trigger 1000s of updates of endpoints.

Comment (howardjohn) [reply pkg/workloadapi/workload.proto:90]: I agree, Kuat. I'll remove from this PR and we can rebuild it differently later

File: pkg/workloadapi/workload.proto
Comment (hzxuzhonghu) [reply pkg/workloadapi/workload.proto:28]: +1        without a doc, it is hard to know what it does.

**Comment by costinm** [comment]
> The l4 subset of Istio rbac

You mean the internal envoy representation ( envoy RBAC ) ? To be honest I would use Istio proto - or k8s proto for NetworkPolicy if we decide to reuse/extend. But Kuat had a great point that it shouldn't be sent per endpoint, in particular since it's applied server side.

We will use additional resources in ztunnel - ProxyConfig, whatever policy, maybe more.

What still has me confused is the difference between service VIP and pod-to-pod IPs. Do we want to mix them in a single resource ?
In a way I like the concept - it's an IP to IPInfo, both may have multiple addresses.

**Comment by howardjohn** [comment]
> > The l4 subset of Istio rbac
> 
> You mean the internal envoy representation ( envoy RBAC ) ? To be honest I would use Istio proto - or k8s proto for NetworkPolicy if we decide to reuse/extend. But Kuat had a great point that it shouldn't be sent per endpoint, in particular since it's applied server side.
> 
> We will use additional resources in ztunnel - ProxyConfig, whatever policy, maybe more.
> 
> What still has me confused is the difference between service VIP and pod-to-pod IPs. Do we want to mix them in a single resource ? In a way I like the concept - it's an IP to IPInfo, both may have multiple addresses.

I have gone back and forth on service embedded... my thinking was pod changes way more than  service. If we embed, a pod changes just changes 1 workload (100b or so).

If we don't, we need to push the Service for all services associated. Unless we have some xds magic, that would mean resending the entire list of IPs for the service, quite a bit more than 100b?

**Comment by kyessenov** [comment]
Embedding VIP in pod IP also means that every time a service changes, you have to resend all the pod IP immutable data like SA, namespace, etc. I tend to think this is better handled by sending the VIP association separately. What this could be is a mapping from IP -> union of pod metadata, service port mapping, pods

**Comment by kyessenov** [comment]
@howardjohn IP is 32 bits. An address on a modern machine is bigger than that :)

File: pkg/workloadapi/workload.proto
Comment (costinm) [inline pkg/workloadapi/workload.proto:71]: Why ? RBAC is applied on the server side, not client - and has selectors.

We should use a separate type, no mixing.

**Review by costinm — APPROVED** [review]
One more rbac comment that can be addressed in a separate PR, we'll refine the proto anyways.

Rest seems reasonable and safe enough for ambient branch.

File: pkg/workloadapi/workload.proto
Comment (howardjohn) [reply pkg/workloadapi/workload.proto:71]: I think this is old commit, I deleted?

Comment (howardjohn) [reply pkg/workloadapi/workload.proto:71]: oops, forgot to push

**Review by stevenctl — APPROVED** [review]

**Comment by costinm** [comment]
See my doc on Service behavior, we can discuss it there.

We should not duplicate EDS - but if we send the list of pods, and each pod
has the list of services (small, fixed) - then all we need is a reverse
map that can be easily created by the client, and you have service -> IP.

We only get updates when pods go away and start - and that will update the
relevant services.

**Comment by costinm** [comment]
I would name the field 'Address' - size can be 4, 16, 6 or 18 bytes. ( last
2 if we decide to embed the port ). The main goal is to not parse ascii
that much, it is meant to be fast and memory friendly.

**Pull request merged**
